File: Dynamic_Programming/c++/06_가장긴증가하는부분수열.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/11053
// 2

#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int N;
vector<int> arr;
int dp[1001];

int main() {
	cin >> N;
	for (int i = 0; i < N; i++) {
		int tmp;
		cin >> tmp;
		arr.push_back(tmp);
		dp[i] = 1;
	}

	int result = 1;
	for (int i = 1; i < N; i++) {
		for (int j = 0; j < i; j++) {
			if (arr[j] < arr[i])
				dp[i] = max(dp[i], dp[j] + 1);
		}
		result = max(result, dp[i]);
	}
	
	cout << result;
}

File: Dynamic_Programming/python/06_가장긴증가하는부분수열.py
# fast campus 강의
# https://www.acmicpc.net/problem/11053
# 1

n = int(input())
array = list(map(int, input().split()))
dp = [1] * n
for i in range(1, n):
    for j in range(0, i):
        if array[j] < array[i]:
            dp[i] = max(dp[i], dp[j] + 1)
print(max(dp))